Client/Frontend/Widgets/PhotonActionSheetProtocol.swift

```diff
@@ -0,0 +1,200 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Shared
+import Storage
+
+protocol PhotonActionSheetProtocol {
+    var tabManager: TabManager { get }
+    var profile: Profile { get }
+}
+
+extension PhotonActionSheetProtocol {
+    typealias PresentableVC = UIViewController & UIPopoverPresentationControllerDelegate
+    
+    func presentSheetWith(actions: [[PhotonActionSheetItem]], on viewController: PresentableVC, from view: UIView) {
+        let sheet = PhotonActionSheet(actions: actions)
+        sheet.modalPresentationStyle =  UIDevice.current.userInterfaceIdiom == .pad ? .popover : .overCurrentContext
+        sheet.photonTransitionDelegate = PhotonActionSheetAnimator()
+        
+        if let popoverVC = sheet.popoverPresentationController {
+            popoverVC.backgroundColor = UIColor.clear
+            popoverVC.delegate = viewController
+            popoverVC.sourceView = view
+            popoverVC.sourceRect = CGRect(x: view.frame.width/2, y: view.frame.size.height * 0.75, width: 1, height: 1)
+            popoverVC.permittedArrowDirections = UIPopoverArrowDirection.up
+        }
+        viewController.present(sheet, animated: true, completion: nil)
+    }
+    
+    //Returns a list of actions which is used to build a menu
+    //parameter OpenURL is a clousre that can open a given URL in some view controller. It is up to the class using the menu to know how to open the url
+    func getHomePanelActions(openURL: @escaping (URL) -> Void, vcDelegate: PageOptionsVC) -> [PhotonActionSheetItem] {
+        let openQR = PhotonActionSheetItem(title: "QR Scanner", iconString: "menu-ScanQRCode") { action in
+            let qrCodeViewController = QRCodeViewController()
+            qrCodeViewController.qrCodeDelegate = vcDelegate
+            let controller = UINavigationController(rootViewController: qrCodeViewController)
+            vcDelegate.present(controller, animated: true, completion: nil)
+        }
+        
+        let openSettings = PhotonActionSheetItem(title: "Settings", iconString: "menu-Settings") { action in
+            let settingsTableViewController = AppSettingsTableViewController()
+            settingsTableViewController.profile = self.profile
+            settingsTableViewController.tabManager = self.tabManager
+            settingsTableViewController.settingsDelegate = vcDelegate
+            
+            let controller = SettingsNavigationController(rootViewController: settingsTableViewController)
+            controller.popoverDelegate = vcDelegate
+            controller.modalPresentationStyle = UIModalPresentationStyle.formSheet
+            vcDelegate.present(controller, animated: true, completion: nil)
+        }
+        
+        let openTopSites = PhotonActionSheetItem(title: Strings.AppMenuTopSitesTitleString, iconString: "menu-panel-TopSites") { action in
```

Comment:
I see what is happening here, and I'm not sure I like it. 

Putting the actual action with the menu item seems appealing to start with, but quickly becomes difficult when you want to perform the same action without going through the menu. e.g. `findInPage`, `bookmark` can be triggered from multiple places, and you want the identical code to be triggered for each. 

This was part of the rationale of trying to get a redux pattern in place.